propolis-server-lib/src/server.rs

```diff
@@ -0,0 +1,431 @@
+                                ErrorKind::InvalidData,
+                                "Cannot parse disk path",
+                            )
+                        })?;
+                        let bdf: pci::Bdf =
+                            dev.get("pci-path").ok_or_else(|| {
+                                Error::new(
+                                    ErrorKind::InvalidData,
+                                    "Cannot parse disk PCI",
+                                )
+                            })?;
+                        init.initialize_block(&chipset, path, bdf)?;
+                    }
+                    "pci-virtio-viona" => {
+                        let name = dev.get_string("vnic").ok_or_else(|| {
+                            Error::new(
+                                ErrorKind::InvalidData,
+                                "Cannot parse vnic name",
+                            )
+                        })?;
+                        let bdf: pci::Bdf =
+                            dev.get("pci-path").ok_or_else(|| {
+                                Error::new(
+                                    ErrorKind::InvalidData,
+                                    "Cannot parse vnic PCI",
+                                )
+                            })?;
+                        init.initialize_vnic(&chipset, name, bdf)?;
+                    }
+                    _ => {
+                        return Err(Error::new(
+                            ErrorKind::InvalidData,
+                            format!("Unknown driver in config: {}", driver),
+                        ));
+                    }
+                }
+            }
+
+            // Finalize device.
+            chipset.device().pci_finalize(&disp.ctx());
+            init.initialize_fwcfg(&chipset, properties.vcpus)?;
+            init.initialize_cpus()?;
+            Ok(())
+        })
+        .map_err(|err| {
+            HttpError::for_internal_error(format!(
+                "Failed to initialize machine: {}",
+                err.to_string()
+            ))
+        })?;
+
+    let (tx, rx) = watch::channel(StateChange {
+        gen: 0,
+        state: propolis::instance::State::Initialize,
+    });
+    instance.print();
+
+    let id = properties.id;
+    instance.on_transition(Box::new(move |next_state| {
+        println!("state cb: {:?}", next_state);
+        let last = (*tx.borrow()).clone();
+        let _ = tx.send(StateChange { gen: last.gen + 1, state: next_state });
+
+        // If we transition to the destroyed state, we should be able to
+        // free up resources for the underlying VM.
+        if matches!(next_state, propolis::instance::State::Destroy) {
+            let r = propolis::vmm::destroy_vm(id.to_string());
+            if let Err(e) = r {
+                eprintln!("Failed to destroy VM ({}): {}", id, e);
+            } else {
+                eprintln!("Destroyed VM ({})", id);
+            }
+        }
```

Comment:
This bit is new, and worth taking a look at.

Rather than leaving defunct VMs sitting around, the server now explicitly destroys them. WDYT?

Reply:
Destroying it rather than leaving it dangling sounds good.  Do you see a natural way for this to fit into the propolis-lib API, rather than reaching all the way in for the vmm method and ID string?

Reply:
Done - I moved this into the state machine driver in `instance.rs`.

This now requires that the invocation of `VmmHdl::destroy` act on "&self", instead of "&mut self". I'll toy around with managing the object more exclusively (perhaps behind an `Option`?). I'd really like to ensure that after destroy is called, we won't try to communicate with the VM.